Example fuzzer for RT-N12 B1 httpd

libumem opened the pull request:
## Checklist

### Which kind of PR do you create?

- [x] This PR only contains minor fixes.
- [ ] This PR contains major feature update.
- [ ] This PR introduces a new function/api for Qiling Framework.

### Coding convention?

- [x] The new code conforms to Qiling Framework naming convention.
- [x] The imports are arranged properly.
- [x] Essential comments are added.
- [ ] The reference of the new code is pointed out.

### Extra tests?

- [x] No extra tests are needed for this PR.
- [ ] I have added enough tests for this PR.
- [ ] Tests will be added after some discussion and review.

### Changelog?

- [x] This PR doesn't need to update Changelog.
- [ ] Changelog will be updated after some proper review.
- [ ] Changelog has been updated in my PR.

### Target branch?

- [x] The target branch is dev branch.

### One last thing

- [x] I have read the [contribution guide](https://docs.qiling.io/en/latest/contribution/)

-----

elicn commented on examples/fuzzing/rt_n12/fuzz.py:
```diff
@@ -0,0 +1,173 @@
+from qiling.extensions.coverage import utils as cov_utils
+import argparse
+
+payload_location = 0 # location in snapshot to be modified
```
Global variables are not encouraged in general, but if you still need them please try to minimize their scope: it looks like `fuzz` function is the right scope for it.

libumem replied:
>  it looks like fuzz function is the right scope for it.

Moved.

elicn commented on examples/fuzzing/rt_n12/fuzz.py:
```diff
@@ -0,0 +1,173 @@
+    used to create this snapshot
+    """
+    params = ql.os.resolve_fcall_params({'sockfd': INT,'buf':POINTER, 'len':SIZE_T, 'flags':INT })
+    ql.log.info(buf)
```
`buf` is not defined.
Did you mean `params[buf]`?

libumem replied:
> Did you mean params[buf]?

Yes

> Also consider a meaningful debug message and not just emitting the value out.

Done

elicn replied:
I had a typo; that should have been `params['buf']`, of course.

libumem replied:
Fixed.

elicn commented on examples/fuzzing/rt_n12/fuzz.py:
```diff
@@ -0,0 +1,173 @@
+    """
+    params = ql.os.resolve_fcall_params({'sockfd': INT,'buf':POINTER, 'len':SIZE_T, 'flags':INT })
+    ql.log.info(buf)
+    return
```
No need to use an empty `return` statement at the end of the function.

libumem replied:
Removed

elicn commented on examples/fuzzing/rt_n12/fuzz.py:
```diff
@@ -0,0 +1,173 @@
+    def start_afl(ql: Qiling):
+        """Have Unicorn fork and start instrumentation.
+        """
+        avoids = []
```
The following seems to be more elegant:
```python
avoids = [
    0x0040577c,  # 501 not implemented, avoid wasting time on non POST/HEAD/GET/etc requests
    0x00405b98,  # jump back to select() loop?
    0x004059b4   # same as above?
]
```

libumem replied:
Fixed

elicn commented on examples/fuzzing/rt_n12/fuzz.py:
```diff
@@ -0,0 +1,173 @@
+    ql.restore(snapshot="httpd.bin")
+    global payload_location
+    payload_location = ql.mem.search(b'GET /FUZZME HTTP/1.1')[0]
+    ql.log.info("Location of input data")
```
Replace with:
```python
ql.log.info(f"Location of input data: {payload_location:#010x}")
```

libumem replied:
Done

elicn commented on examples/fuzzing/rt_n12/fuzz.py:
```diff
@@ -0,0 +1,173 @@
+    ql.log.info("Location of input data")
+    ql.log.info(hex(payload_location))
+    ql.os.set_api('send', my_send, QL_INTERCEPT.CALL) # avoid crash due to the socket no longer being valid on clent end
+    ql.hook_address(start_afl, 0x00404b20) # kick off from the start of snapshot
```
The `0x00404b20` is magic number that keeps showing up here. Better define it as a variable somewhere with a proper comment that says where it comes from.

libumem replied:
Agreed.

elicn commented on examples/fuzzing/rt_n12/fuzz.py:
```diff
@@ -0,0 +1,173 @@
+    def my_nvram_get(self, ql:Qiling):
+        self.ql.log.info("NVRAM get call")
+        key_addr = self.ql.arch.regs.read("A0")
+        key = str(ql.mem.string(key_addr))
```
`mem.string` is deprecated, please use `os.utils.read_cstring` instead.

libumem replied:
Done

elicn commented on examples/fuzzing/rt_n12/fuzz.py:
```diff
@@ -0,0 +1,173 @@
+        self.ql.log.info("NVRAM get call")
+        key_addr = self.ql.arch.regs.read("A0")
+        key = str(ql.mem.string(key_addr))
+        if self.dbg_level == QL_VERBOSE.DEBUG:
```
Would it make more sense to just use the `debug` logging level instead of checking for it explicitly?
```python
self.ql.log.debug(f'key: {key}')
```

libumem replied:
Done

elicn commented on examples/fuzzing/rt_n12/fuzz.py:
```diff
@@ -0,0 +1,173 @@
+        if self.dbg_level == QL_VERBOSE.DEBUG:
+            self.ql.log.info('value ' + val)
+
+        self.ql.mem.write(self.nvram_addr, bytes(val[::-1], 'utf-8'))
```
A comment would help to understand why you are writing these bytes backwards.

libumem replied:
Done

elicn commented on examples/fuzzing/rt_n12/fuzz.py:
```diff
@@ -0,0 +1,173 @@
+        self.fake_return() # emulate return
+        return
+    def populate_nvram(self):
+        f = open('nvram', 'rb')
```
The file is opened but never closed. Better to use the file context manager instead:
```python
with open('nvram', 'rb') as f:
    # do everything here...
```

libumem replied:
Done

elicn commented on examples/fuzzing/rt_n12/fuzz.py:
```diff
@@ -0,0 +1,173 @@
+        self.ql = Qiling(self.cmdline,rootfs=self.rootfs, verbose=dbg_level)
+        self.nvram = {}
+        self.nvram_addr = self.ql.mem.map_anywhere(size=4096, info='nvram label')
+        self.ql.fast_mode=True
```
What is this?

libumem replied:
Added comment addressing that.

elicn replied:
I meant, what is `fast_mode`? It looks like something you added and never used.

libumem replied:
Removed.

elicn commented on examples/fuzzing/rt_n12/fuzz.py:
```diff
@@ -0,0 +1,173 @@
+    if args.run:
+        with cov_utils.collect_coverage(emu.ql, 'drcov', 'output.cov'):
+            emu.ql.run()
+    if args.fuzz and args.filename:
```
It is not clear whether `run` and `fuzz` are mutually exclusive or not. If they are, then it needs to be reflected through arg parser.

libumem replied:
Fixed.

elicn reviewed (changes requested):
Hi there and thanks for the contribution; more examples help people to moderate their learning curve.
For that very reason it is imperative that the examples will be as clean as possible and at the same time heavily commented. What are you doing and why you are doing things they way you do is key for other people to learn from the example. If too many things remain mysterious or unclear people will not learn too much, or worse, it even might confuse them.

Will appreciate you taking the time to go over the requested changes and make the necessary adjustments. Also here are a few general comments:
 1. Try to adhere to PEP8 conventions, it really helps to keep the code clean and tidy
 2. Use proper annotations also for return values, not only for arguments
 3. Add more comments explaining what you are doing; things that are obvious to you might not be for others
 4. I am not sure we are allowed to attach the firmware files here. If you have a direct like to the firmware on the vendor's website, that would be great

Cheers!

libumem commented:
>  Try to adhere to PEP8 conventions, it really helps to keep the code clean and tidy

I've since run a linter to address this. 

> Use proper annotations also for return values, not only for arguments

Should be fixed now.

> Add more comments explaining what you are doing; things that are obvious to you might not be for others

Added multiple comment sections to aid understanding. 

> I am not sure we are allowed to attach the firmware files here. If you have a direct like to the firmware on the vendor's website, that would be great

Removed firmware file (which was the wrong one anyway) and added a link to the manfacturer's page.

elicn commented:
Great, thanks for making the changes.
Please see the two remaining comments.

libumem commented:
@elicn Addressed the feedback you gave. Much appreciated.

libumem commented:
Will someone else please approve this?

xwings commented:
Hi, 

Welcome to Qiling Framework.

Nice work. Thanks!

Sorry for the delay.
